nemo/core/neural_graph/graph_outputs.py

```diff
@@ -0,0 +1,301 @@
+
+        # Ok, set thee "manual" output.
+        self._manual_outputs[key] = GraphOutput(value.ntype, value.producer_step_module_port)
+
+    def __getitem__(self, key: str) -> GraphOutput:
+        """
+            Returns the bound output associated with the given key.
+            Uses default or manual dict depending whether there are some manual outputs or not.
+
+            Args:
+                key: Name of the bound input.
+        """
+        if len(self._manual_outputs) > 0:
+            return self._manual_outputs[key]
+        else:  # Use default dict.
+            return self._default_outputs[key]
+
+    def __delitem__(self, key: str):
+        """
+            Raises:
+                TypeError as deletion of a bound input port is not allowed.
+        """
+        raise TypeError("Deleting a bound output is not allowed")
+
+    def __iter__(self):
+        """
+            Returns:
+                Iterator over the outputs - depending whether there are some manual outputs or not.
+        """
+        if len(self._manual_outputs) > 0:
+            return iter(self._manual_outputs)
+        else:  # Use default dict.
+            return iter(self._default_outputs)
+
+    def __len__(self) -> int:
+        """
+            Returns:
+                The number of outputs - depending whether there are some manual outputs or not.
+        """
+        if len(self._manual_outputs) > 0:
+            return len(self._manual_outputs)
+        else:  # Use default dict.
+            return len(self._default_outputs)
+
+    def bind(self, tensors_ref: List[NmTensor], port_names: Optional[str] = None):
+        """
+            Binds the "default" outputs.
+
+            Args:
+                tensors_ref: List of tensors to be added.
+                port_names: List of port names (visible outside). If None: using internal tensor "output port names".
+        """
+        # Set names.
+        if port_names is None:
+            port_names = [tensor.name for tensor in tensors_ref]
+
+        for name, tensor in zip(port_names, tensors_ref):
+            # Check the presence of the port name in "default" dictionary.
+            if name in self._default_outputs.keys():
+                # Name present - use the name being combination of producer and port names.
+                name = (
+                    str(tensor.producer_step_number) + "_" + tensor.producer_name + "_" + tensor.name
+                )  # last = port name
+
+                logging.warning(
+                    "Setting unigue name of the default output port `{}` produced in step {} by `{}` to `{}`".format(
+                        tensor.name, tensor.producer_step_number, tensor.producer_name, name
+                    )
+                )
+            # Still, "overwrite" it.
+            self._default_outputs[name] = GraphOutput(tensor.ntype, tensor.producer_step_module_port)
```

Comment:
Why do we overwrite the name? We have tensor.unique_name, why not use it?

Reply:
Because it relies on uuid which is specific for a given process. Next time you run your script - uuid will be **different**. 

What happens in here relates to binding ("recording") of "default" outputs (tensors). They are stored in a dictionary, where key = module output port name.

However, if you plug the same module twice in your graph, we do not want to override the already bound tensor. So for new one I generate "truly unique name" (step_producer_port) that can be exported/imported/nested.

Reply:
Ps. there is a _reason_ why e.g. layers in keras have a *name* property...

Reply:
If you use the same module twice, it should generate a new unique NmTensor, so I do not understand the argument for plugging the same module twice.

Is there an issue that arises when you import and export a graph? I fail to see why `Next time you run your script - uuid will be different.` is an issue.